WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift

```diff
@@ -944,124 +939,93 @@ class AbstractPostListViewController: UIViewController,
-            tableView.reloadRows(at: [indexPath], with: .fade)
-        }
-
-        let postService = PostService(managedObjectContext: ContextManager.sharedInstance().mainContext)
+        let originalStatus = apost.status
 
         let trashed = (apost.status == .trash)
 
-        postService.trashPost(apost, success: {
-            // If we permanently deleted the post
+        let repository = PostRepository(coreDataStack: ContextManager.shared)
+        do {
+            try await repository.trash(TaggedManagedObjectID(apost))
             if trashed {
                 PostCoordinator.shared.cancelAnyPendingSaveOf(post: apost)
                 MediaCoordinator.shared.cancelUploadOfAllMedia(for: apost)
             }
-        }, failure: { [weak self] (error) in
 
-            guard let strongSelf = self else {
-                return
-            }
-
-            if let error = error as NSError?, error.code == type(of: strongSelf).HTTPErrorCodeForbidden {
-                strongSelf.promptForPassword()
+            let message = NSLocalizedString("Post moved to trash.", comment: "A short message explaining that a post was moved to the trash bin.")
+            let undoAction = NSLocalizedString("Undo", comment: "The title of an 'undo' button. Tapping the button moves a trashed post out of the trash folder.")
+            let notice = Notice(title: message, actionTitle: undoAction, actionHandler: { [weak self] accepted in
+                if accepted {
+                    self?.restorePost(apost, toStatus: originalStatus ?? .draft)
+                }
+            })
+            ActionDispatcher.dispatch(NoticeAction.dismiss)
+            ActionDispatcher.dispatch(NoticeAction.post(notice))
+        } catch {
+            if let error = error as NSError?, error.code == type(of: self).HTTPErrorCodeForbidden {
+                promptForPassword()
             } else {
                 WPError.showXMLRPCErrorAlert(error)
             }
 
-            if let index = strongSelf.recentlyTrashedPostObjectIDs.firstIndex(of: postObjectID) {
-                strongSelf.recentlyTrashedPostObjectIDs.remove(at: index)
-                // We don't really know what happened here, why did the request fail?
-                // Maybe we could not delete the post or maybe the post was already deleted
-                // It is safer to re fetch the results than to reload that specific row
-                DispatchQueue.main.async {
-                    strongSelf.updateAndPerformFetchRequestRefreshingResults()
-                }
+            // We don't really know what happened here, why did the request fail?
+            // Maybe we could not delete the post or maybe the post was already deleted
+            // It is safer to re fetch the results than to reload that specific row
+            DispatchQueue.main.async {
+                self.updateAndPerformFetchRequestRefreshingResults()
             }
-        })
+        }
```

Comment:
Have you considered moving some of the logic here into a ViewModel-like object, so that we can test it?

One point of friction I can see is that there is no such object for this VC that I can see at the moment. Another is that this is a base class, meaning that the actual subclasses the app uses at runtime might need additional behavior and might be left with having to create their own ViewModel subclasses. This would be confusing. At the same time, it would be a sign of an abstraction that no longer serves us well and we might want to update its design.

Reply:
That'd be much better than this view controller's current state. But I don't want to just move this small part (trashing, deleting, restoring) out, leaving this view controller in a weird state with mixed-patterns. A large refactor on this screen also seems unnecessary at this point.

Reply:
> leaving this view controller in a weird state with mixed-patterns

Good point 🤔 

I am a fan of _sprouting_ (a la [Working Effectively with Legacy Code](https://www.amazon.com/Working-Effectively-Legacy-Michael-Feathers/dp/0131177052)) better code from existing code, even if what is added/extracted is a little portion of the total. But ☝️ having mixed patterns is a concern.

I guess it would be acceptable only if we had an agreed upon expectation that "all new and modified code should follow <pattern>." That is not the case at the moment.